neps/nep-0413.md

```diff
@@ -0,0 +1,153 @@
+
+### How to Return the Signed Message in a Safe Way
+Sending the signature in a query string to an arbitrary URL (even within the correct domain) is not secure as the data can be leaked (e.g. through headers, etc). Using URL fragments instead will improve security, since the fragment isn't included in the Referrer. Further constraining the callback URL to a fixed value will improve security even more.
+
+### NEAR Signatures
+NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+## Specification
+Given the previous [rationales](#rationale-and-alternatives), this NEP proposes the wallets to implement a `verifyOwner` method following the specifications bellow.
+
+
+### Goal
+The `verifyOwner` method must take a plain text `message` and transform it into a signature.
+
+### Input Interface
+In order to sign a message, `verifyOwner` must implement the following input interface:
+
+```jsx
+interface VerifyOwnerParams {
+  message: string; // The message requested to be signed, should contain the app name.
+  callbackUrl?: string; // Applicable to browser wallets (e.g. MyNearWallet). This is the callback url once the signing is approved. Defaults to `window.location.href`.
+  meta?: string; // Applicable to browser wallets (e.g. MyNearWallet) extra data that will be passed to the callback url once the signing is approved.
+}
+```
+
+### Structure and Signature
+In order to respect all the points layed in [rationales](#rationale-and-alternatives), `verifyOwner` must:
+
+1. Embed the passed `message` into a predefined structure (see below).
+2. Stringify the structure 
+2. Compute the `SHA256 hash` of the *stringified structure*.
+3. Use a full-access key to sign the resulting `SHA256 hash`.
+
+#### Structure
+The structure must have the following interface:
+
+```rust
+struct Payload {
+  accountId: string; // Mandatory: the account name as plain text (e.g. "alice.near")
+  message: string; // The same message passed in `VerifyOwnerParams.message` 
+  blockId: string; // The hash of a block created close to the time of signature
+  publicKey: string; // public counterpart of the key used to sign, encoded as a base64 string (see bellow)
+}
+```
+
+#### Signature
+In order to create a signature, the `Payload` must be converted into its `string` representation (respecting the attribute's order given above), and then hashed using `SHA256`.
+
+For example, assuming that:
+1. `alice.near` wants to sign the message `"berryclub.io"`.
+2. The wallet stores a private key which public counterpart is: `ed25519:6TupyNrcHGTt5XRLmHTc2KGaiSbjhQi1KHtCXTgbcr4Y`, which in base64 is: `ZWQyNTUxOTo2VHVweU5yY0hHVHQ1WFJMbUhUYzJLR2FpU2JqaFFpMUtIdENYVGdiY3I0WQ==`
```

Comment:
Don't see the point in additional base64 encoding of the PublicKey representation of `ed25519:6TupyNrcHGTt5XRLmHTc2KGaiSbjhQi1KHtCXTgbcr4Y`.

To achieve this the **wallet** would to double base64 encode the public key and once:

```js
const publicKeyStr = publicKey.toString() // ed25519:6TupyNrcHGTt5XRLmHTc2KGaiSbjhQi1KHtCXTgbcr4Y 
const publiucKeyStrBytes = decodeUtf8(publicKeyStr); // [12, 13, 133, 137, ...] 
const verifyOwnerPublicKey = encodeBase64(publiucKeyStrBytes); // ZWQyNTUxOTo2VHVweU5yY0hHVHQ1WFJMbUhUYzJLR2FpU2JqaFFpMUtIdENYVGdiY3I0WQ==
```

While the application would need to perform also redundant decoding

```ts
const result = await wallet.verifyOwner(...); 

const verifyOwnerPublicKey = result.publicKey // ZWQyNTUxOTo2VHVweU5yY0hHVHQ1WFJMbUhUYzJLR2FpU2JqaFFpMUtIdENYVGdiY3I0WQ==
const publiucKeyStrBytes = decodeBase64(result.publicKey); // [12, 13, 133, 137, ...]
const publicKeyStr = encodeUtf8(publiucKeyStrBytes) // ed25519:6TupyNrcHGTt5XRLmHTc2KGaiSbjhQi1KHtCXTgbcr4Y 
const publicKey = PublicKey.fromString(publicKeyStr)
```

Instead of all this, propose that the encoding remains to the native Near representation of `"ed25519:6TupyNrcHGTt5XRLmHTc2KGaiSbjhQi1KHtCXTgbcr4Y"`.

This is also how the keys are encoded in the `account.getAccessKeys()`, it will be easier to perform verification of the validity of the public key.

Reply:
Can we be sure that **all** keys (assuming new ones are added) will have a string format that can be encoded in simple text? Because using `base64` we **ensure** that any key can be sent back as an URL parameter.

I honestly don't know how standardized are key-formats, for which I was assuming there might be some that have "weird" characters.

Reply:
@gagdiez 
> Can we be sure that all keys (assuming new ones are added) will have a string format that can be encoded in simple text? Because using base64 we ensure that any key can be sent back as an URL parameter.

Yup, we can be sure. Every key can be represented by the key type (curve) and the bytes (parameters) that are used for that key. Since bytes will be encoded into `base64` we are sure that it is safe to be put into URL parameter. Just the `:` colon character would need to be encoded (`encodeURIComponent(publicKey.toString()`) when this is returned by URL param.

> I honestly don't know how standardized are key-formats, for which I was assuming there might be some that have "weird" characters.

- From the sign-in response from MyNearWallet I get the following parameters 
  - account_id=tote.testnet
  - all_keys=ed25519%3A4uAW6WoYKVLzvaVwniBMPUujSxJi7fVDs5bVFjb3j
- It is how the `PublicKey.toString()` exports with: https://github.com/near/near-api-js/blob/96785cb/packages/near-api-js/src/utils/key_pair.ts#L56-L58)
- The response from `view_access_key_list` is resolving like that: https://docs.near.org/api/rpc/access-keys#view-access-key-list 
- It is how it is stored also: https://github.com/near/NEPs/blob/17af66e32a83e0bfea4e8110b59902daf09e7ab4/specs/GenesisConfig/StateRecord.md#public_key

Not sure if there is formal proposal for the key-format, but seems that `<key-type>:<base-64-key-bytes>` is the format that is widely accepted.

Reply:
I'm in favour of using the public string directly as it appears consistent with existing implementations.

Reply:
> Not sure if there is formal proposal for the key-format, but seems that `<key-type>:<base-64-key-bytes>` is the format that is widely accepted.

Yep- think if this is the standard, and its the same as what is in the response for `view_access_key_list` - this would be the best way forward. Especially seeing that the key is already base64 encoded.